File: VRSampleApp/VRSampleApp/Classes/PresentationLayer/Modules/CustomVideoModule/View/Controller/CustomVideoViewController.m
//
//  CustomVideoViewController.m
//  VRSampleApp
//
//  Copyright © 2017 Ooyala Inc. All rights reserved.
//

#import "CustomVideoViewController.h"
#import "CustomVideoTransitionManager.h"

@interface CustomVideoViewController () <UITextFieldDelegate, UIGestureRecognizerDelegate>

@property (weak, nonatomic) IBOutlet UITextField *pCodeTextField;
@property (weak, nonatomic) IBOutlet UITextField *embedCodeTextField;
@property (weak, nonatomic) IBOutlet UIButton *openButton;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *containerCenterYConstraint;
@property (nonatomic) CustomVideoTransitionManager *transitionManager;

extern CGFloat kContentViewCornerRadius;
extern CGFloat kDefaultSpacing;

@end


@implementation CustomVideoViewController

#pragma mark - Constants

CGFloat kContentViewCornerRadius = 8.0;
CGFloat kDefaultSpacing = 8.0;

#pragma mark - Life cycle

- (void)viewDidLoad {
  [super viewDidLoad];
  
  [self configureObjects];
  [self configureUI];
}

#pragma mark - Private properties

- (void)configureObjects {
  // Add keyboard notifications
  [NSNotificationCenter.defaultCenter addObserver:self
                                         selector:@selector(keyboardNotification:)
                                             name:UIKeyboardWillShowNotification
                                           object:NULL];
  
  [NSNotificationCenter.defaultCenter addObserver:self
                                         selector:@selector(keyboardNotification:)
                                             name:UIKeyboardWillHideNotification
                                           object:NULL];
}

- (void)configureUI {
  // Transition
  self.transitionManager = [CustomVideoTransitionManager new];
  self.transitioningDelegate = self.transitionManager;
  
  // Container view
  self.containerView.layer.masksToBounds = YES;
  self.containerView.layer.cornerRadius = kContentViewCornerRadius;
}

- (void)updateUIWithKeyboardActionWithKeyboardFrame:(CGRect)keyboardFrame isShowing:(BOOL)isShowing {
  if (!isShowing) {
    self.containerCenterYConstraint.constant = 0;
    return;
  }
  
  if (self.pCodeTextField.isEditing) {
    [self updateContainerViewPostionWithPCodeTestFieldEditingWithKeyboardFrame:keyboardFrame];
  } else {
    [self updateContainerViewPostionWithKeyboardFrame:keyboardFrame andView:self.containerView];
  }
}

- (void)updateContainerViewPostionWithPCodeTestFieldEditingWithKeyboardFrame:(CGRect)keyboardFrame {
  CGRect convertedEmbedCodeTextFieldFrame = [self.embedCodeTextField convertRect:self.embedCodeTextField.bounds
                                                                          toView:self.view];
  CGFloat convertedEmbedCodeTextFieldMaxY = CGRectGetMaxY(convertedEmbedCodeTextFieldFrame);
  CGFloat delta = convertedEmbedCodeTextFieldMaxY + kDefaultSpacing - keyboardFrame.origin.y;
  
  if (delta > 0) {
    if (fabs(self.view.bounds.size.height - keyboardFrame.size.height) > self.containerView.bounds.size.height) {
      [self updateContainerViewPostionWithKeyboardFrame:keyboardFrame andView:self.containerView];
    } else {
      self.containerCenterYConstraint.constant = -(self.view.bounds.size.height / 2 - self.containerView.bounds.size.height / 2);
    }
  } else if (delta < 0 && CGRectGetMinY(self.containerView.frame) < 0) {
    self.containerCenterYConstraint.constant = 0;
  }
}

- (void)updateContainerViewPostionWithKeyboardFrame:(CGRect)keyboardFrame andView:(UIView *)view {
  CGRect convertedViewFrame = [view convertRect:view.bounds toView:self.view];
  CGFloat containerBottomMaxY = CGRectGetMaxY(convertedViewFrame);
  CGFloat delta = containerBottomMaxY + kDefaultSpacing - keyboardFrame.origin.y;
  
  self.containerCenterYConstraint.constant += -delta;
}

#pragma mark - Actions

- (IBAction)pCodeTextFieldDidChange:(UITextField *)sender {
  [self.viewModel pCodeDidUpdated:sender.text];
}

- (IBAction)embedCodeTextFieldDidChange:(UITextField *)sender {
  [self.viewModel embedCodeDidUpdated:sender.text];
}

- (IBAction)openButtonAction:(UIButton *)sender {
  [self.viewModel openButtonDidTapped];
}

- (IBAction)cancelButtonAction:(UIButton *)sender {
  [self dismissViewControllerAnimated:YES completion:NULL];
}

- (IBAction)tapOnViewGestureAction:(UITapGestureRecognizer *)sender {
  [self.pCodeTextField endEditing:YES];
  [self.embedCodeTextField endEditing:YES];
}

#pragma mark - Notifications

- (void)keyboardNotification:(NSNotification *)notification {
  BOOL isShowing = notification.name == UIKeyboardWillShowNotification;
  NSDictionary *userInfo = notification.userInfo;
  CGRect endFrame = [(NSValue *)userInfo[UIKeyboardFrameEndUserInfoKey] CGRectValue];
  NSTimeInterval duration = [userInfo[UIKeyboardAnimationDurationUserInfoKey] doubleValue]; // or 0
  UIViewAnimationOptions animationCurve = UIViewAnimationOptionCurveEaseInOut;

  [self updateUIWithKeyboardActionWithKeyboardFrame:endFrame isShowing:isShowing];
  
  [UIView animateWithDuration:duration delay:0 options:animationCurve animations:^{
    [self.view layoutIfNeeded];
  } completion:NULL];
}

#pragma mark - CustomVideoViewModelDelegate

- (void)openButtonEnabledChanged:(BOOL)enabled {
  [self.openButton setEnabled:enabled];
}

- (void)dismisCurrentModule:(void (^)(BOOL finished))completion {
  [self.pCodeTextField endEditing:YES];
  [self.embedCodeTextField endEditing:YES];
  
  [self dismissViewControllerAnimated:YES completion:^{
    completion(YES);
  }];
}

#pragma mark - UITextFieldDelegate

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
  if (textField == self.pCodeTextField) {
    [self.embedCodeTextField becomeFirstResponder];
  } else {
    [self.pCodeTextField becomeFirstResponder];
  }
  
  return YES;
}

#pragma mark - UIGestureRecognizerDelegate

- (BOOL)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldReceiveTouch:(UITouch *)touch {
  return !CGRectContainsPoint(self.containerView.frame, [touch locationInView:self.view]);
}

@end

File: SwiftVRSkinSampleApp/SwiftVRSkinSampleApp/Classes/PresentationLayer/Modules/CustomVideoModule/View/Controller/CustomVideoViewController.swift
//
//  CustomVideoViewController.swift
//  SwiftVRSkinSampleApp
//
//  Copyright © 2017 Ooyala, Inc. All rights reserved.
//

import UIKit


class CustomVideoViewController: UIViewController {
  
  // MARK: - Constants
  
  private enum Constants {
    static let contentViewCornerRadius: CGFloat = 8.0
    static let defaultSpacing: CGFloat = 8.0
  }
 
  // MARK: - Public properties
  
  var viewModel: CustomVideoViewModel!
  
  // MARK: - IBOutlets

  @IBOutlet weak var containerView: UIView!
  @IBOutlet weak var pCodeTextField: UITextField!
  @IBOutlet weak var embedCodeTextField: UITextField!
  @IBOutlet weak var openButton: UIButton!
  @IBOutlet weak var containerCenterYConstraint: NSLayoutConstraint!
  
  // MARK: - Private properties
  
  private var transitionManager: CustomVideoTransitionManager!
  
  // MARK: - Init/deinit
  
  deinit {
    NotificationCenter.default.removeObserver(self)
  }
  
  // MARK: - Life cycle
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    configureObjects()
    configureUI()
  }
  
  // MARK: - Private functions
  
  private func configureObjects() {
    
    // Add keyboard notifications
    
    NotificationCenter
      .default
      .addObserver(self, selector: #selector(keyboardNotification(notification:)), name: .UIKeyboardWillShow, object: nil)
    
    NotificationCenter
      .default
      .addObserver(self, selector: #selector(keyboardNotification(notification:)), name: .UIKeyboardWillHide, object: nil)
  }
  
  private func configureUI() {
    
    // Transition
    
    transitionManager = CustomVideoTransitionManager()
    transitioningDelegate = transitionManager
    
    // Container view
    
    containerView.layer.masksToBounds = true
    containerView.layer.cornerRadius = Constants.contentViewCornerRadius
  }
  
  private func updateUIWithKeyboardAction(keyboardFrame: CGRect, isShowing: Bool) {
    if !isShowing {
      containerCenterYConstraint.constant = 0
      return
    }
    
    if pCodeTextField.isEditing {
      updateContainerViewPostionWithPCodeTestFieldEditing(keyboardFrame: keyboardFrame)
    } else if embedCodeTextField.isEditing {
      updateContainerViewPostion(with: keyboardFrame, view: containerView)
    }
  }
  
  private func updateContainerViewPostionWithPCodeTestFieldEditing(keyboardFrame: CGRect) {
    let convertedEmbedCodeTextFieldFrame = embedCodeTextField.convert(embedCodeTextField.bounds, to: self.view)
    let convertedEmbedCodeTextFieldMaxY = convertedEmbedCodeTextFieldFrame.maxY
    
    let delta = convertedEmbedCodeTextFieldMaxY + Constants.defaultSpacing - keyboardFrame.origin.y
    
    if delta > 0 {
      if abs(self.view.bounds.size.height - keyboardFrame.size.height) > containerView.bounds.size.height {
        updateContainerViewPostion(with: keyboardFrame, view: containerView)
      } else {
        containerCenterYConstraint.constant = -(self.view.bounds.height / 2 - containerView.bounds.height / 2)
      }
    } else if delta < 0 && containerView.frame.minY < 0 {
      containerCenterYConstraint.constant = delta
    }
  }
  
  private func updateContainerViewPostion(with keyboardFrame: CGRect, view: UIView) {
    let convertedViewFrame = view.convert(view.bounds, to: self.view)
    let containerBottomMaxY = convertedViewFrame.maxY
    let delta = containerBottomMaxY + Constants.defaultSpacing - keyboardFrame.origin.y

    containerCenterYConstraint.constant += -delta
  }
  
  // MARK: - Actions
  
  @IBAction func pCodeTextFieldDidChange(_ sender: UITextField) {
    viewModel.pCodeDidUpdated(pCode: sender.text)
  }
  
  @IBAction func embedCodeTextFieldDidChange(_ sender: UITextField) {
    viewModel.embedCodeDidUpdated(embedCode: sender.text)
  }
  
  @IBAction func openButtonAction(_ sender: Any) {
    viewModel.openButtonDidTapped()
  }
  
  @IBAction func cancelButtonAction(_ sender: Any) {
    dismiss(animated: true, completion: nil)
  }
  
  @IBAction func tapOnViewGestureAction(_ sender: UITapGestureRecognizer) {
    pCodeTextField.endEditing(true)
    embedCodeTextField.endEditing(true)
  }
  
  // MARK: - Notifications

  @objc private func keyboardNotification(notification: Notification) {
    let isShowing = notification.name == .UIKeyboardWillShow

    if let userInfo = notification.userInfo {
      guard let endFrame = (userInfo[UIKeyboardFrameEndUserInfoKey] as? NSValue)?.cgRectValue else {
        return
      }
      
      let duration: TimeInterval = (userInfo[UIKeyboardAnimationDurationUserInfoKey] as? NSNumber)?.doubleValue ?? 0
      let animationCurveRawNSN = userInfo[UIKeyboardAnimationCurveUserInfoKey] as? NSNumber
      let animationCurveRaw = animationCurveRawNSN?.uintValue ?? UIViewAnimationOptions.curveEaseInOut.rawValue
      let animationCurve: UIViewAnimationOptions = UIViewAnimationOptions(rawValue: animationCurveRaw)
      
      updateUIWithKeyboardAction(keyboardFrame: endFrame, isShowing: isShowing)
      
      UIView.animate(withDuration: duration,
                     delay: 0,
                     options: animationCurve,
                     animations: { self.view.layoutIfNeeded() },
                     completion: nil)
    }
  }
  

}

// MARK: - CustomVideoViewModelDelegate

extension CustomVideoViewController: CustomVideoViewModelDelegate {
  
  func openButtonEnabledChanged(enabled: Bool) {
    openButton.isEnabled = enabled
  }
  
  func dismisCurrentModule(completion: ((_ finished: Bool) -> Void)?) {
    pCodeTextField.endEditing(true)
    embedCodeTextField.endEditing(true)
    
    dismiss(animated: true) {
      completion?(true)
    }
  }
  
  
}

// MARK: - UITextFieldDelegate

extension CustomVideoViewController: UITextFieldDelegate {

  func textFieldShouldReturn(_ textField: UITextField) -> Bool {
    if textField == pCodeTextField {
      embedCodeTextField.becomeFirstResponder()
    } else {
      pCodeTextField.becomeFirstResponder()
    }
    
    return true
  }
  
  
}

// MARK: - UIGestureRecognizerDelegate

extension CustomVideoViewController: UIGestureRecognizerDelegate {
  
  func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldReceive touch: UITouch) -> Bool {
    return !containerView.frame.contains(touch.location(in: view))
  }
  
  
}
